### PR TITLE
Add cancellations to data transfer operations

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
There is more in plot manager etc - next time.

### CODE REVIEW

File: src/Common/Core/Impl/Extensions/StreamExtensions.cs [hunk 1]
```diff
@@ -3,25 +3,26 @@
...
 using System;
 using System.IO;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.Common.Core {
     public static class StreamExtensions {
-        public static async Task CopyAndFlushAsync(this Stream source, Stream destination, IProgress<long> progress = null) {
-                await source.CopyToAsync(destination, progress);
+        public static async Task CopyAndFlushAsync(this Stream source, Stream destination, CancellationToken cancellationToken, IProgress<long> progress = null) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: Usually in public methods it is a last parameter and has a default value `default(CancellationToken)`. Do we want to follow this pattern?

File: src/Common/Core/Impl/Extensions/StreamExtensions.cs [hunk 2]
```diff
@@ -3,25 +3,26 @@
...
+                await source.CopyToAsync(destination, cancellationToken, progress);
                 await destination.FlushAsync();
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: You can pass `cancellationToken` here as well

File: src/Common/Core/Impl/Extensions/StreamExtensions.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: I know, but there are two more defaults and I actually wanted to make ct required.

File: src/Common/Core/Impl/Extensions/StreamExtensions.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: OK

File: src/Common/Core/Impl/Extensions/StreamExtensions.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: Ok. What about `CopyToAsync`? It doesn't have defaults.

Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: Sure, makes sense

Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/StreamExtensions.cs]: Actually, I'd rather remove default `null` value for `progress` here. By default, we want to provide feedback about the long operation, and user should explicitly opt-out from that feedback.

File: src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs [hunk 3]
```diff
@@ -73,7 +73,7 @@ public class RemoteUriWebService : IRemoteUriWebService {
                         SetResponseHeaders(response, context.Response, localBaseUrl, remoteBaseUrl);
                         using (Stream respStream = response.GetResponseStream())
                         using (Stream outStream = context.Response.OutputStream) {
-                            await respStream.CopyAndFlushAsync(outStream);
+                            await respStream.CopyAndFlushAsync(outStream, ct);
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs]: Async call inside finally block? @int19h , @karthiknadig , is it really required?

Comment (MikhailArkhipov) [inline src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs]: Sure thing

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**